File: tsp.rb
start = Time.now

universities = {
    'BTH' => {"lat" => 56.18, "long" => 15.59},
    'Uppsala' => {"lat" => 59.85, "long" => 17.63},
    'Lund' => {"lat" => 55.71, "long" => 13.20},
    'KTH' => {"lat" => 59.34, "long" => 18.07},
    'Chalmers' => {"lat" => 57.68, "long" => 11.97},
    'Luleå' => {"lat" => 65.61, "long" => 22.14},
    'Umeå' => {"lat" => 63.82, "long" => 20.30},
    'Linköping' => {"lat" => 58.39, "long" => 15.57},
    'Karlstad' => {"lat" => 59.40, "long" => 13.58},
    'Örebro' => {"lat" => 59.25, "long" => 15.24},
    'Linné' => {"lat" => 56.85, "long" => 14.82}
}

university_names = universities.keys()
permutations = {}

N = university_names.length
p = [N]*(N + 1)

shortest_trip = 1000000
shortest_trip_universities = {}

i = 0
while i < N
    p[i] = p[i] - 1

    j = i%2 * p[i]

    tmp = university_names[i]
    university_names[i] = university_names[j]
    university_names[j] = tmp

    cities = university_names.clone
    trip_length = 0
    for i in 0..(cities.length - 2)
        lat1 = universities[cities[i + 1]]["lat"]
        lat0 = universities[cities[i]]["lat"]
        long1 = universities[cities[i + 1]]["long"]
        long0 = universities[cities[i]]["long"]
        trip_length += Math.sqrt((lat1 - lat0)*(lat1 - lat0) + (long1 - long0)*(long1 - long0))
    end

    if trip_length < shortest_trip
        shortest_trip = trip_length
        shortest_trip_universities = cities
    end

    i = 1
    while p[i] == 0
        p[i] = i
        i += 1
    end
end



finish = Time.now

puts "Shortest trip length: " + shortest_trip.to_s
puts shortest_trip_universities.join(" --> ")

diff = finish - start
puts "Execution time: " + diff.to_s

File: tsp.c
#include <stdio.h>
#include <limits.h>
#include <math.h>
#include <string.h>

int main () {
    int n = 11;
    const char *universities[11] = {"BTH", "Uppsala", "Lund", "KTH", "Chalmers", "Luleå", "Umeå", "Linköping", "Karlstad", "Örebro", "Linné"};

    char *mut_universities[11] = {"BTH", "Uppsala", "Lund", "KTH", "Chalmers", "Luleå", "Umeå", "Linköping", "Karlstad", "Örebro", "Linné"};

    const double coordinates[11][2] = {{56.18, 15.59}, {59.85, 17.63}, {55.71, 13.20}, {59.34, 18.07}, {57.68, 11.97}, {65.61, 22.14}, {63.82, 20.30}, {58.39, 15.57}, {59.40, 13.58}, {59.25, 15.24}, {56.85, 14.82}};

    int p[12] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};

    double shortest_trip = INT_MAX;
    char *shortest_trip_universities[11] = {"", "", "", "", "", "", "", "", "", "", ""};

    int i = 0;
    while (i < n) {
        p[i]--;

        int j = i%2 * p[i];

        char *tmp = mut_universities[i];
        mut_universities[i] = mut_universities[j];
        mut_universities[j] = tmp;

        double trip_length = 0.0;
        for (int k = 0; k < n - 1; k++) {
            char *city0 = mut_universities[k];
            char *city1 = mut_universities[k + 1];

            int index0 = -1;
            int index1 = -1;
            for (int l = 0; l < n; l++) {
                if (strcmp(universities[l], city0) == 0) {
                    index0 = l;
                }

                if (strcmp(universities[l], city1) == 0) {
                    index1 = l;
                }
            }

            double lat1 = coordinates[index1][0];
            double lat0 = coordinates[index0][0];
            double long1 = coordinates[index1][1];
            double long0 = coordinates[index0][1];

            trip_length += sqrt((lat1 - lat0)*(lat1 - lat0) + (long1 - long0)*(long1 - long0));
        }


        if (trip_length < shortest_trip) {
            shortest_trip = trip_length;
            for (int m = 0; m < n; m++) {
                shortest_trip_universities[m] = mut_universities[m];
            }
        }

        i = 1;
        while (p[i] == 0) {
            p[i] = i;
            i++;
        }
    }

    printf("Shortest trip length: %f\n", shortest_trip);
    for (int i = 0; i < n - 1; i++) {
        printf("%s --> ", shortest_trip_universities[i]);
    }
    printf("%s\n", shortest_trip_universities[n - 1]);
    printf("Execution time: \n");

    return 0;
}
